src/eepromfs.c

```diff
@@ -0,0 +1,933 @@
+ */
+static int eepfs_find_handle(const char * const path)
+{
+    /* Sanity check */
+    if ( path == NULL )
+    {
+        return EEPFS_EBADINPUT;
+    }
+
+    /* Ignore the leading '/' on the path (if present) */
+    const char * normalized_path = path + (path[0] == '/');
+
+    /* Check if the path matches an entry in the #eepfs_files array */
+    for ( size_t i = 1; i < eepfs_files_count; ++i )
+    {
+        if ( strcmp(normalized_path, eepfs_files[i].path) == 0 )
+        {
+            /* A file exists matching that path */
+            return i;
+        }
+    }
+
+    /* File does not exist */
+    return EEPFS_ENOFILE;
+}
+
+/**
+ * @brief Resolves a file handle to a file descriptor.
+ * 
+ * File descriptors can be used for reading or writing.
+ * 
+ * Getting a file repeatedly will always return the same pointer.
+ * This means that the cursor is shared across all usage of the file.
+ * Be sure to reset the cursor when you are done with it!
+ * 
+ * Negative handles indicate an upstream error occurred.
+ * The zero value will never be used as a file handle, even
+ * though it does map to an entry in the `eepfs_files` table.
+ * The signature entry is not intended to be used as a file.
+ * 
+ * @param[in] handle
+ *            The file handle of an entry in the `eepfs_files` table
+ * 
+ * @return A pointer to the file descriptor or NULL if the handle is invalid
+ */
+static eepfs_file_t * const eepfs_get_file(const int handle)
+{
+    /* Check if the handle appears to be valid */
+    if ( handle > 0 && handle < eepfs_files_count )
+    {
+        return &eepfs_files[handle];
+    }
+
+    return NULL;
+}
+
+/**
+ * @brief Reads a specific amount of data from a file starting from the cursor.
+ *
+ * @param[in]  path
+ *             Path of file in EEPROM filesystem to read from
+ * @param[out] dest
+ *             Buffer to read into
+ * @param[in]  len
+ *             Number of bytes to read
+ *
+ * @return The actual number of bytes read or a negative value on failure.
+ */
+static int eepfs_cursor_read(eepfs_file_t * const file, void * const dest, const size_t len)
+{
+    if ( file == NULL )
+    {
+        return EEPFS_EBADHANDLE;
+    }
+
+    if ( dest == NULL ) 
+    {
+        return EEPFS_EBADINPUT;
+    }
+
+    size_t cursor = file->cursor;
+    size_t bytes_to_read = len;
+
+    /* Bounds check to make sure it doesn't read past the end */
+    if ( cursor + bytes_to_read >= file->max_cursor )
+    {
+        /* It will, let's shorten it */
+        bytes_to_read = file->max_cursor - cursor;
+    }
+
+    if ( bytes_to_read <= 0 )
+    {
+        /* No bytes to read; bail early */
+        return 0;
+    }
+
+    /* Calculate the starting point to read out of EEPROM from */
+    size_t current_block = file->start_block + (cursor / EEPROM_BLOCK_SIZE);
+    size_t block_byte_offset = cursor % EEPROM_BLOCK_SIZE;
+    int bytes_read = 0;
+
+    uint8_t eeprom_buf[EEPROM_BLOCK_SIZE];
+    uint8_t * ptr = dest;
```

Comment:
non-const pointers may help with these. I think @rasky's suggestion was to simplify these

Reply:
I borrowed this pattern from `dragonfs.c`:
https://github.com/DragonMinded/libdragon/blob/87bcff52696abe21a1a322a895677d26ce764e46/src/dragonfs.c#L930
https://github.com/DragonMinded/libdragon/blob/87bcff52696abe21a1a322a895677d26ce764e46/src/dragonfs.c#L981-L982

Reply:
I'd love to hear @rasky's thoughts on this.

Reply:
I saw that in `dragonfs.c` and I think it's wrong there as well. I see zero reasons why the pointer `buf` should be const; in fact, I see zero reasons why a function argument should ever declared const, unless you're coming from a different language where you're used to program the type system more and you're trying to borrow the same concepts to C. In general, I think we can agree that in idiomatic C function arguments are rarely/never defined as const, so I think `dragonfs` should be changed as well.

Reply:
It is my opinion that `const` is a useful tool for signaling intent, and there is merit in communicating that an API will not be modifying certain data (specifically pointers).

However, I agree that `void * const` for the `src`/`dest` parameters is worthless, communicates nothing, and necessitates more code just to work around it.

This has been fixed in https://github.com/DragonMinded/libdragon/pull/125/commits/8a71b66c8626ab99272b1a4a3e679bd9db54ceee